Better gemm support for cublas and cpu

cnuernber opened the pull request:
Opening this up for review immediately.

These test pass on cpu and cuda for float32 and float64 datatypes:

https://github.com/tech-ascent/tech.compute/blob/master/src/tech/compute/verify/tensor.clj#L275

tqchen commented:
Please request reviews from the reviewers and other contributors

liangfu commented on src/contrib/cublas/cublas_utils.h:
```diff
@@ -0,0 +1,53 @@
+  do {                                          \
+    int error = static_cast<int>(fn);                      \
+    CHECK_EQ(error, CUBLAS_STATUS_SUCCESS) << "CUBLAS: " << get_cublas_error_string(error); \
+  } while (0);
```
How about remove the semicolon at the end, since we would typically call the macro with another semicolon.

liangfu commented on src/contrib/cublas/cublas_utils.h:
```diff
@@ -0,0 +1,53 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
```
Change the year to 2018 ?

liangfu commented on src/contrib/cublas/cublas_utils.h:
```diff
@@ -0,0 +1,53 @@
+namespace tvm {
+namespace contrib {
+
+inline const char* get_cublas_error_string(int error) {
```
+1 for handling cublas error string with inline function instead of macro.

liangfu replied:
However, can we change the name of the function using camel cases, following [Google C/C++ style](https://google.github.io/styleguide/cppguide.html#Function_Names) ?

cnuernber commented:
@tqchen:  I cannot request reviews via the normal gear next to the reviewers menu heading; I think this is an issue with project settings and my github handle's relationship with the project.  

I am still working on this, step today is to handle transposed (i.e. strides are reversed from normal) and make sure the smaller of the 2 strides is 1.

@liangfu Sounds good, all good suggestions.

cnuernber commented:
@liangfu:  ready for next review.

tqchen commented:
@cnuernber you can just request reviewers by at them, Thanks!

cnuernber commented:
@masahi: Here is the PR from the thread if you would like to review.

tqchen commented on include/tvm/contrib/gemm.h:
```diff
@@ -0,0 +1,94 @@
+
+  using namespace runtime;
+
+  inline int columnStride(DLTensor* tensor) {
```
some note about style:

- do not put auxiliary function header in inlcude, these features are better suited in the private part, so put them in src/contrib/gemm

- Use CamelCase for functions(google C style)

tqchen commented on include/tvm/contrib/gemm.h:
```diff
@@ -0,0 +1,94 @@
+    // If the tensor itself is transposed then it will have strides
+    // backward from what we expect.  Regardless, the max of the strides
+    // (the other stride is 1) is the column stride.
+    if (tensor->strides)
```
style, enclose with {} if you can

cnuernber commented:
@tqchen:  I modified the code.  
1.  Proper camel case and braces - done.
2.  Moving utility functions out of main header - done with caveats:

I moved the utility functions into a details header similar to what boost does.  Moving them to a cc file would involve adding the file *and* modifying the cmake system to include 1 cc file.  This seems like a more heavyweight change especially for essentially 1 liners that someone who was auditing the gemm code would want to read; thus in a header near the gemm code makes the system, in my opinion, more discoverable.

I am trying to make the result as readable as possible and avoid modifying the build system for those short one liners.  Is there a more appropriate place to put these sort of helper functions than a separate cc file?  

Please consider that the CallGemm function itself is a private utility function for each gemm implementation that has to be described in a header file as it is a template operator-based function.

In any case, I am happy to make more changes and ready for next review step.

tqchen commented:
@cnuernber What I mean by the private header is not about moving them to cc file, but instead put it under src/contrib/cblas/gemm_common.h, then you can include the header by using a relative path to the file. 

There are other examples of private headers, e.g. https://github.com/dmlc/tvm/blob/master/src/lang/attr_functor.h

The general idea is that these headers can be used by TVM internally, but we do not intend to expose it to the users of TVM

cnuernber commented:
@tqchen:  Got it, sorry for the misunderstanding.  Updated with your recommentation.

Anything else you would like I am happy to do :-).

tqchen commented:
@liangfu can you take another look and https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

liangfu commented on src/contrib/cublas/cublas_utils.h:
```diff
@@ -0,0 +1,53 @@
+  do {                                          \
+    int error = static_cast<int>(fn);                      \
+    CHECK_EQ(error, CUBLAS_STATUS_SUCCESS) << "CUBLAS: " << getCublasErrorString(error); \
+  } while (0)  // ; intentially left off.
```
actually, i'm favor of removing the comment in this line.

cnuernber commented:
@liangfu: Addressed :-).  Ready for next round.

tqchen commented:
Thanks, @cnuernber for contributing and @liangfu for the reviews, this is now merged
